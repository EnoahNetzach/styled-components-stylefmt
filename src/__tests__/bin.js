import { spawn } from 'child_process'
import fs from 'fs-extra'
import globby from 'globby'
import os from 'os'
import path from 'path'

const tmpdir = fs.mkdtempSync(os.tmpdir())

const copyTestFile = (file) => {
  const tmpFile = path.join(tmpdir, path.basename(file))

  fs.copy(file, tmpFile)

  return tmpFile
}

const run = args =>
  new Promise((resolve) => {
    const bin = spawn('node', [path.join(__dirname, '../../bin/index.js')].concat(args), {
      stdio: 'pipe',
    })

    let error = ''
    let output = ''

    bin.stdout.on('data', (data) => {
      output += data.toString()
    })

    bin.stderr.on('data', (data) => {
      error += data.toString()
    })

    bin.on('close', code => resolve({ code, error, output }))
  })

describe('bin', () => {
  let args = []

  describe('no args', () => {
    test('prints the help page by default', async () => {
      const { code, error, output } = await run(args)

      expect(code).toBe(0)
      expect(error).toBe('')
      expect(output).not.toBe('')
      expect(output).toMatchSnapshot()
    })
  })

  describe('help', () => {
    beforeEach(() => {
      args = ['--help']
    })

    test('prints the help page', async () => {
      const { code, error, output } = await run(args)

      expect(code).toBe(0)
      expect(error).toBe('')
      expect(output).not.toBe('')
      expect(output).toMatchSnapshot()
    })
  })

  describe('diff', () => {
    beforeEach(() => {
      args = ['--diff']
    })

    test('prints the diff for a single file', async () => {
      const inputFile = path.join(__dirname, './fixtures/hard/invalid-indentation.actual.js')
      const { code, error, output } = await run([inputFile, ...args])

      expect(code).toBe(0)
      expect(error).toBe('')
      expect(output).not.toBe('')
      expect(output).toMatchSnapshot()
    })

    test('prints the diff for multiple files', async () => {
      const inputGlob = path.join(__dirname, './fixtures/**/*.actual.js')
      const { code, error, output } = await run(['--recursive', inputGlob, ...args])

      expect(code).toBe(0)
      expect(error).toBe('')
      expect(output).not.toBe('')
      expect(output).toMatchSnapshot()
    })
  })

  describe('write', () => {
    beforeEach(() => {
      args = []
    })

    afterEach(() => {
      fs.removeSync(tmpdir)
      fs.mkdirsSync(tmpdir)
    })

    test('formats a single file and prints to stdout', async () => {
      const inputFile = copyTestFile(path.join(__dirname, './fixtures/hard/invalid-indentation.actual.js'))
      const { code, error, output } = await run([inputFile, ...args])

      expect(code).toBe(0)
      expect(error).toBe('')
      expect(output).not.toBe('')
      expect(output).toMatchSnapshot()
    })

    test('formats a single file and overwrites it', async () => {
      const fixturePath = path.join(__dirname, './fixtures/hard/invalid-indentation.actual.js')
      const inputFile = copyTestFile(fixturePath)
      const { code, error, output } = await run([inputFile, inputFile, ...args])

      expect(code).toBe(0)
      expect(error).toBe('')
      expect(output).toBe('')
      expect(fs.readFileSync(fixturePath)).not.toEqual(fs.readFileSync(inputFile))
    })

    test("formats a single file and doesn't overwrite it if it's unchanged", async () => {
      const fixturePath = path.join(__dirname, './fixtures/real-world/Circle.actual.js')
      const inputFile = copyTestFile(fixturePath)
      const { code, error, output } = await run([inputFile, inputFile, ...args])
      const fixture = fs.readFileSync(fixturePath)

      expect(code).toBe(0)
      expect(error).toBe('')
      expect(output).toEqual(fixture.toString())
      expect(fixture).toEqual(fs.readFileSync(inputFile))
    })

    test('formats multiple files and overwrites them', async () => {
      const fixtureGlob = path.join(__dirname, './fixtures/hard/*.actual.js')
      const inputGlob = path.join((await globby([fixtureGlob])).map(copyTestFile).map(path.dirname)[0], '*.js')
      const { code, error, output } = await run(['--recursive', inputGlob, ...args])

      expect(code).toBe(0)
      expect(error).toBe('')
      expect(output).toMatch(/3 files are formatted\./)
    })
  })
})
