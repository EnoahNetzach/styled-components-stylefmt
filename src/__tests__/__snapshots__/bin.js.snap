// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bin diff prints the diff for a single file 1`] = `
"src/__tests__/fixtures/hard/invalid-indentation.actual.js
import styled from 'styled-components'

// ⚠ 2 indentation errors ⚠
const Comp = () => {
  const Button = styled.button\`
    color: blue;
    background: red;
    display: block;
  \`

  return Button
}

// ⚠ 2 indentation errors ⚠
const Comp2 = () => {
  const InnerComp = () => {
    const Button = styled.button\`
      color: blue;
      background: red;
      display: block;
    \`

    return Button
  }

  return InnerComp()
}

"
`;

exports[`bin diff prints the diff for multiple files 1`] = `
"src/__tests__/fixtures/hard/indentation.actual.js
import styled, { keyframes } from 'styled-components'

// None of the below should throw indentation errors
const Comp = () => {
  const Button = styled.button\`
    color: blue;
  \`

  return Button
}

const Comp2 = () => {
  const InnerComp = () => {
    const Button = styled.button\`
      color: blue;
    \`

    return Button
  }

  return InnerComp()
}

const Button = styled.button\`
  color: blue;
\`

const animations = {
  spinnerCircle: keyframes\`
    0% {
      opacity: 0;
    }

    100% {
      opacity: 1;
    }
  \`,
}


src/__tests__/fixtures/hard/invalid-indentation.actual.js
import styled from 'styled-components'

// ⚠ 2 indentation errors ⚠
const Comp = () => {
  const Button = styled.button\`
    color: blue;
    background: red;
    display: block;
  \`

  return Button
}

// ⚠ 2 indentation errors ⚠
const Comp2 = () => {
  const InnerComp = () => {
    const Button = styled.button\`
      color: blue;
      background: red;
      display: block;
    \`

    return Button
  }

  return InnerComp()
}


src/__tests__/fixtures/hard/source-maps.actual.js
import styled from 'styled-components'

// ⚠️ BAD INDENTATION at 5:1 ⚠️
const Button = styled.button\`
  color: blue;
\`

// Correct example
const Button2 = styled.button\`
  color: blue;
\`

// ⚠️ BAD INDENTATION at 10:5 ⚠️
const Button3 = styled.button\`
  color: blue;
\`

// ⚠️ BAD INDENTATION at 22:5 ⚠️
const Button4 = styled.button\`
  color: blue;
  background: \${color};
  display: block;
\`

// ⚠️ BAD INDENTATION at 28:5 ⚠️
const Button5 = styled.button\`
  color: blue;
  background: \${color};
  display: block;
\`

// ⚠️ BAD INDENTATION at 35:5 ⚠️
const Button6 = styled.button\`
  color: blue;
  \${\`
      background: red;
  \`}
  display: block;
\`


src/__tests__/fixtures/interpolations/complex.actual.js
There is no difference with the original file.

src/__tests__/fixtures/interpolations/invalid.actual.js
import styled from 'styled-components'

const color = 'red'

// ⚠️ BAD INDENTATION ⚠️
const Button2 = styled.button\`
  display: block;
  color: \${color};
  background: blue;
\`


src/__tests__/fixtures/interpolations/valid.actual.js
There is no difference with the original file.

src/__tests__/fixtures/real-world/Circle.actual.js
There is no difference with the original file.

src/__tests__/fixtures/simple/global.actual.js
// Global variables

// ⚠️ Wrong indentation ⚠️
const Button = styled.div\`
  color: red;
\`

// ⚠️ Wrong indentation ⚠️
const animation = keyframes\`
  0% {
    opacity: 0;
  }
\`

// ⚠️ Wrong indentation ⚠️
injectGlobal\`
  html {
    margin: 0;
  }
\`

const styles = css\`
  color: blue;
\`


src/__tests__/fixtures/simple/helpers.actual.js
import styled, { css, keyframes, injectGlobal } from 'styled-components'

/**
 * Valid
 */
const styles = css\`
  color: blue;
\`

const animation = keyframes\`
  0% {
    opacity: 1;
  }

  100% {
    opacity: 0;
  }
\`

const Button = styled.button\`
  \${styles}
  animation: 3s \${animation};
\`

injectGlobal\`
  html {
    margin: 0;
    padding: 0;
  }
\`

// ⚠ Indentation
const styles2 = css\`
  color: blue;
\`

// ⚠ Indentation
const animation2 = keyframes\`
  0% {
    opacity: 1;
  }

  100% {
    opacity: 0;
  }
\`

// ⚠ Indentation
injectGlobal\`
  html {
    margin: 0;
    padding: 0;
  }
\`


src/__tests__/fixtures/simple/imports.actual.js
import notStyled, { css as notCss, keyframes as notKeyframes, injectGlobal as notInjectGlobal } from 'styled-components'

// ⚠️ BAD INDENTATION ⚠️
const Button2 = notStyled.button\`
  color: blue;
\`

const styles = notCss\`
  color: blue;
\`

const animation = notKeyframes\`
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
\`

notInjectGlobal\`
  html {
    color: blue;
  }
\`


src/__tests__/fixtures/simple/invalid.actual.js
import styled from 'styled-components'

// ⚠️ EMPTY BLOCK ⚠️
const Button = styled.button\`

\`

// ⚠️ BAD INDENTATION ⚠️
const Button2 = styled.button\`
  color: blue;
\`


src/__tests__/fixtures/simple/nesting.actual.js
There is no difference with the original file.

src/__tests__/fixtures/simple/valid.actual.js
There is no difference with the original file.
"
`;

exports[`bin help prints the help page 1`] = `
"Usage: styled-components-stylefmt [options] input-name [output-name]

Options:

  -b, --config-basedir   Path to the directory that relative paths defining \\"extends\\"
  -c, --config           Path to a specific configuration file (JSON, YAML, or CommonJS)
  -d, --diff             Output diff against original file
  -r, --recursive        Format list of space seperated files(globs) in place
  -v, --version          Output the version number
  -h, --help             Output usage information
  -i, --ignore-path      Path to a file containing patterns that describe files to ignore.
"
`;

exports[`bin no args prints the help page by default 1`] = `
"Usage: styled-components-stylefmt [options] input-name [output-name]

Options:

  -b, --config-basedir   Path to the directory that relative paths defining \\"extends\\"
  -c, --config           Path to a specific configuration file (JSON, YAML, or CommonJS)
  -d, --diff             Output diff against original file
  -r, --recursive        Format list of space seperated files(globs) in place
  -v, --version          Output the version number
  -h, --help             Output usage information
  -i, --ignore-path      Path to a file containing patterns that describe files to ignore.
"
`;

exports[`bin write formats a single file and prints to stdout 1`] = `
"import styled from 'styled-components'

// ⚠ 2 indentation errors ⚠
const Comp = () => {
  const Button = styled.button\`
    color: blue;
    background: red;
    display: block;
  \`

  return Button
}

// ⚠ 2 indentation errors ⚠
const Comp2 = () => {
  const InnerComp = () => {
    const Button = styled.button\`
      color: blue;
      background: red;
      display: block;
    \`

    return Button
  }

  return InnerComp()
}
"
`;
